### ISSUE TITLE
Please make position and file in location message type optional

### ISSUE THREAD

**Issue by mschuchard** [steelbrain/linter#1447]
I have several linter packages that I made where there are situations where no position or position and file information is available. I need to be able to do something like this:

```javascript
location: {
  file: file,
},

location: {},
```

to be able to upgrade to the v2 API. Alternatively, for the situation where no file information is available, there is directory information available sometimes. What kind of of consequences would result from the following:

```javascript
location: {
  file: dir
},
```

**Comment by steelbrain** [comment]
> What kind of of consequences would result from the following

It'll highlight on tree view correctly but when people click on it, we'll be trying to do `atom.workspace.open(dir)` and that's bad, we don't wanna do that

> I have several linter packages that I made where there are situations where no position or position and file information is available

There are linters that returns positions most of the times, but when something goes horribly wrong they don't output a position. In that kinda scenarios, if the problem is with installation or something the user can control, they should throw an error so linter can show the error notification, otherwise they could just do `position: [[0, 0], [0, 1]]`. This would highlight the first character of the file, highlight the file in tree view as well as show the error in Linter Panel!

I've disallowed location-less messages on v2 based on a very strong personal opinion, they are still supported in v1. Mostly because it's abused, and people who don't use a panel have ZERO way of knowing it exists. I personally program with the panel closed and use inline tooltips and highlights with tree view to help me be productive.

**Comment by mschuchard** [comment]
So if I have a situation where a message is returned for a directory, what would you suggest I do for the v2 API? Is `file` required inside `location` too, or is just `position` that is required?

**Comment by steelbrain** [comment]
There can be multiple files opened at once, and in case of Project scoped linters, they are not even executed on a text editor at all. So yes, it's required.

Can you please share an example of such an error where exact file path is not available? I still think we can throw an error notification for those 😛

**Comment by mschuchard** [comment]
We could still display a message for that situation, except that it would be misleading since it would point to a file that probably does not actually have the error.

**Comment by mschuchard** [comment]
By the way:

> when people click on it, we'll be trying to do atom.workspace.open(dir) and that's bad, we don't wanna do that

does not seem to be happening, so that is good.
